### DISCUSSION TITLE
Sockets and Crazyflie API

### DISCUSSION THREAD

**Discussion by JamesDvp30** [bitcraze/discussions#999] [Q&A]
hi!, now im doing the question with good bases. 

Well im doing a comunication from my pc (windows) to the VM using sockets, i provide an IP address and a port for that, well im sending the info but i see that there is a problem. If i only send 1 command it works, but when i try to send 2 or 3 more i see that the drone simply take off and land. i found that it sends very fast the commands and they arrive together, so i would like to know if there are some advices or recommendations that you can give me.

Here is the part where i send the commands : https://drive.google.com/file/d/18Fr_aIpNBQGd0Cbbd9Da7MSoZ5eCyIxr/view?usp=sharing

Here is the part where i recieve them in the VM: https://drive.google.com/file/d/154B7detI4kMzIOceg3sJ11TR0Qp-luue/view?usp=sharing

i saw that there is a delay too when you debug the code about like 8 segs (green flash on the dongle)

this is part of the code

```python
import logging
from sqlite3 import Time
import time
import socket
import cflib.crtp
from cflib.crazyflie import Crazyflie
from cflib.crazyflie.syncCrazyflie import SyncCrazyflie
from cflib.positioning.motion_commander import MotionCommander
from cflib.utils import uri_helper

URI = uri_helper.uri_from_env(default='radio://0/80/2M/E7E7E7E7E7')

# Only output errors from the logging framework
logging.basicConfig(level=logging.ERROR)

def manejar_comandos():
    server_socket = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
    server_socket.bind(('0.0.0.0', 5000))
    server_socket.listen(1)
    print("server socket active. Esperando comandos...")

    while True:
        client_socket, client_address = server_socket.accept()
        print("Client connected from:", client_address)

        with SyncCrazyflie(URI, cf=Crazyflie(rw_cache='./cache')) as scf:
            mc = MotionCommander(scf)
            mc.take_off()
            print("DRONE TOOK OFF!")
            time.sleep(2)

        while True:
            command_bytes = client_socket.recv(1024)
            if not command_bytes:
                break  #if there is no data it goes out

            command = command_bytes.decode('utf-8')
            print(f"recieved command: {command}")

            if command == "forward":
                print("Ejecutando comando 'forward'")
                # it goes forward
                mc.forward(0.5)
                time.sleep(10)
            elif command == "left":
                print("Ejecutando comando 'left'")
                # Realiza la acción correspondiente, por ejemplo, mc.left(0.5)
                time.sleep(4)
            elif command == "up":
                print("Ejecutando comando 'up'")
                # Realiza la acción correspondiente, por ejemplo, 
                mc.up(0.5)
                time.sleep(4)
            elif command == "down":
                print("Ejecutando comando 'down'")
                # it goes down 
                mc.down(0.3)
                time.sleep(4)
            else:
                print("unreconized command")
                print("Im going to land")
                mc.land()
                break

        print("Client disconnected.")
        print("Im Going to land")
        mc.land()

if __name__ == '__main__':
        cflib.crtp.init_drivers()
        manejar_comandos()
```

**Comment by krichardsson** [comment]
I added some formatting to the code in your question to make it more readable.

The first thing I notice is that it looks like the `while True:` section should be indented one more step as the connection to the CF will be closed when the `SyncCrazyflie` instance goes out of scope. You probably intended this to run while the CF is connected?

In your video it looks like you get into the "unreconized command" part of your code which indicates that there is a problem in your socket communication. Unfortunately this is out of scope for what we can help you with, but maybe a good approach would be to make sure the socket stuff works first. Perhaps write a script without any Crazyflie related code, just to simplify things, and add the flying later when the communication works.

**Reply by JamesDvp30** [reply]
hi!, yes i saw that the problem was on the comunication, due to fast command transmission. if i apply a delay there it works nice. but another thing i would like to know is, does the SyncCrazyflie instance can keep the drone active flying the whole time?(like just waiting),  moreover, i know that if you dont specify mc.takeoff() it does it by itself, but does landing too?

i will keep doing tests,  and let you know if i improve something

**Reply by krichardsson** [reply]
> i would like to know is, does the SyncCrazyflie instance can keep the drone active flying the whole time?

The `SyncCrazyflie` class is only handling the connection to the Crazyflie and does not have any functionality for taking off or flying. The underlying API is asynchronous (with callbacks) which makes it a bit complex to use and the purpose of the `SyncCrazyflie` class is to hide most of this. When using it the way you do (using the `with` keyword) the connection will be setup on the `with` line and closed when it goes out of scope.

```python 
with SyncCrazyflie(URI, cf=Crazyflie(rw_cache='./cache')) as scf:
    # We are now connected and can use the scf object to communicate with the CF
    doing_stuff_that_uses_the_connection(scf)
    # This is where the "with" clause ends and the connection is closed

# The scf object can not be used any more here
```

> i know that if you dont specify mc.takeoff() it does it by itself, but does landing too?

Yes and no. The purpose of the `MotionCommander` class is similar to the `SyncCrazyflie` class in that it hides complexity and presents a cleaner API with higher level functionality. If you use it in a `with` clause it will take off automatically but not if you don't

```python 
with SyncCrazyflie(URI, cf=Crazyflie(rw_cache='./cache')) as scf:
    # We are now connected and can use the scf object to communicate with the CF
    with MotionCommander(scf) as mc:
        # Automatic take off on the line above
        control_the_cf_using_the_mc_object(mc)
        # End of the "with" clause, the motion commander will land the CF
    
    # You can also use the motion commander without a "with" clause, but then you have to take off and land with explicit calls
    # This code should do the same thing as above
    mc2 = MotionCommander(scf)
    mc2.take_off()
    control_the_cf_using_the_mc_object(mc2)
    mc2.land()

    # This is where the outer "with" clause ends and the connection is closed
```

Under the hood python is calling the functions [`__enter__()`](https://github.com/bitcraze/crazyflie-lib-python/blob/c1c42f7a8b58bc95d660247d5a2a6832755dae90/cflib/positioning/motion_commander.py#L125-L127) and [`__exit__()`](https://github.com/bitcraze/crazyflie-lib-python/blob/c1c42f7a8b58bc95d660247d5a2a6832755dae90/cflib/positioning/motion_commander.py#L129-L130) when using the `with` keyword and as you can see all the `MotionCommander` is doing is to call `take_off()` and `land()`

**Comment by JamesDvp30** [comment]
Hi guys! im back. Fortunately i solved that and  here i will add part of the code to help people that wants to work with sockets and threading with the drone.

``` python
import socket
import time
from queue import Queue
import cflib.crtp
from cflib.crazyflie import Crazyflie
from cflib.crazyflie.syncCrazyflie import SyncCrazyflie
from cflib.positioning.motion_commander import MotionCommander
from cflib.utils import uri_helper
import threading

# Configura el URI del Crazyflie
URI = uri_helper.uri_from_env(default='radio://0/80/2M/E7E7E7E7E7')
comandos_queue = Queue()

# Configurar el socket
server_socket = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
server_socket.bind(('0.0.0.0', 5000))
server_socket.listen(1)
print("Servidor socket activo. Esperando comandos...")

def manejar_comandos():
    try:
        with SyncCrazyflie(URI, cf=Crazyflie(rw_cache='./cache')) as scf:
            mc = MotionCommander(scf)
            mc.take_off()
            print("Im Connected")
            while True:
                # Obtener el próximo comando de la cola
                command = comandos_queue.get()
                
                # Procesar y ejecutar el comando
                if command == "neutral":
                    print("No hago nada")
                    mc.stop()

                elif command == "right":
                    mc.right(0.5)
                    print("Going right")
                elif command == "left":
                    mc.left(0.5)
                    print(" going left")
                elif command == "lift":
                    mc.up(0.3)
                    print("going up")
                else:
                    mc.land()
                    print("landing")
    except Exception as e:
        print("landing...")
        mc.land()



def manejar_conexion():
    while True:
        # Aceptar una conexión
        client_socket, client_address = server_socket.accept()
        print("Cliente conectado desde:", client_address)

        while True:
            # Recibir el comando
            command_bytes = client_socket.recv(1024)
            if not command_bytes:
                break  # Si no se reciben datos, salimos del bucle interno

            command = command_bytes.decode('utf-8')
            print(f"Comando recibido: {command}")

            # Añadir el comando a la cola
            comandos_queue.put(command)

        print("Cliente desconectado.")
        client_socket.close()

# Iniciar los hilos

if __name__ == '__main__':
    cflib.crtp.init_drivers()
    threading.Thread(target=manejar_conexion).start()
    threading.Thread(target=manejar_comandos).start()
```
Here is a idea about how to keep your drone alive flying waiting for a command. (there are few thing to improve btw)

This is the Server Side of a Socket communication. Remember to make an Client side where you specify an Host IP and a Port and obviously and this part

`word_to_send = variable.encode('utf-8')  
socket.connect((host, port))
socket.sendall(comando)
`
On the client side

Is there an example to land the drone using a key from my keyboard in the discussions? because im planning to use it as emergency land during the presentation of the project.  Thank Youu
